capa/features/extractors/dnfile/insn.py

```diff
@@ -8,58 +8,68 @@
 
 from __future__ import annotations
 
-from typing import TYPE_CHECKING, Dict, Tuple, Iterator, Optional
+from typing import TYPE_CHECKING, Any, Dict, Tuple, Iterator, Optional
 
 if TYPE_CHECKING:
     from dncil.cil.instruction import Instruction
     from dncil.cil.body import CilMethodBody
     from capa.features.common import Feature
 
-from dncil.clr.token import StringToken
+import dnfile
+from dncil.clr.token import StringToken, InvalidToken
 from dncil.cil.opcode import OpCodes
 
 import capa.features.extractors.helpers
 from capa.features.insn import API, Number
-from capa.features.common import String
+from capa.features.common import String, Characteristic
 from capa.features.extractors.dnfile.helpers import (
+    resolve_dotnet_token,
     read_dotnet_user_string,
     get_dotnet_managed_imports,
     get_dotnet_unmanaged_imports,
     get_dotnet_managed_method_names,
 )
 
 
-def get_imports(ctx: Dict) -> Dict:
-    if "imports_cache" not in ctx:
-        ctx["imports_cache"] = {}
-
+def get_managed_imports(ctx: Dict) -> Dict:
+    if "managed_imports_cache" not in ctx:
+        ctx["managed_imports_cache"] = {}
         for (token, name) in get_dotnet_managed_imports(ctx["pe"]):
-            ctx["imports_cache"][token] = name
-        for (token, name) in get_dotnet_unmanaged_imports(ctx["pe"]):
-            ctx["imports_cache"][token] = name
+            ctx["managed_imports_cache"][token] = name
+    return ctx["managed_imports_cache"]
 
-    return ctx["imports_cache"]
+
+def get_unmanaged_imports(ctx: Dict) -> Dict:
+    if "unmanaged_imports_cache" not in ctx:
+        ctx["unmanaged_imports_cache"] = {}
+        for (token, name) in get_dotnet_unmanaged_imports(ctx["pe"]):
+            ctx["unmanaged_imports_cache"][token] = name
+    return ctx["unmanaged_imports_cache"]
 
 
 def get_methods(ctx: Dict) -> Dict:
     if "methods_cache" not in ctx:
         ctx["methods_cache"] = {}
-
         for (token, name) in get_dotnet_managed_method_names(ctx["pe"]):
             ctx["methods_cache"][token] = name
-
     return ctx["methods_cache"]
 
 
+def get_callee_name(ctx: Dict, token: int) -> str:
+    name: str = get_managed_imports(ctx).get(token, "")
+    if not name:
+        name = get_unmanaged_imports(ctx).get(token, "")
+        if not name:
+            name = get_methods(ctx).get(token, "")
```

Comment:
is this order arbitrary? should we document why it is the way it is?

Reply:
order matters; I'll add documentation to make this more clear. thank you!